### ISSUE TITLE
Code organization

### ISSUE THREAD

**Issue by albertcthomas** [ramp-kits/rl_simulator#1]
Summarizing ongoing discussion on code organization. What we have in mind for now:

1.	One folder for each system: `acrobot/`, `inverted_pendulum/`, … with for each system:
-	`env.py`: the gym-like env
-	`generate_initial_trace.py`: to generate static datasets from a given policy (random or other), these static datasets are used to evaluate the generative regressors. As suggested by @kegl a better name would be `generate_static_traces.py`
-	`reward_function.py` (containing the reward as a function of the observations)
-	`problem.py` (the `problem.py` of the generative regressors with CV and workflow)
-	`data/` (`X_train.csv`, `X_test.csv` generated by `generate_initial_trace.py` and `metadata.json`) 
-	`submissions/`
-	`agents/`
2.	One `rlkit` folder for the model based `rlkit` package (to be installed with `pip`). This contains the main RL script to be called and functions common to all systems.
-	`model_based_rl.py`: the main script
-	`model_env.py`: creating the gym-like env from a trained submission.
-	`train_test_split_initial_trace.py`: splitting the generated trace in train and test, to be called in `generate_static_trace.py`
-	`utils.py`: lots of functions… data preprocessing, loading initial trace, creating agent from the model based env, generating data from a given agent. This would need some refactoring.
-	`tests/`

Some parts could use/need some refactoring. I suggest to first have a draft of this organization in the repo, most of the refactoring could be done afterwards.

cc @kegl @gabriel-hurtado

**Comment by albertcthomas** [comment]
To store the trained submission and the traces collected during the rollout the structure would be `submissions/<submission>/<agent>/seed_<i>/epoch_<i>/(trained_submission.pkl, trace.csv`). Trained agents and trained value functions could also be saved at the same location.

An example of the current behavior to make things a bit more clear, and especially with the first epoch (epoch 0) for which we have 2 (very similar) cases: the one where we start from an initial (random) policy collecting data on the system, the one where we start from an initial trace.
1. With the random policy: do the rollout and in `epoch_0/` save `trace.csv` and train the submission from this `trace.csv` giving a `trained_submission.pkl` saved also in `epoch_0/`. Then use this model for epoch 1
2. With a given initial trace: put this trace in `epoch_0/` and train the submission from it giving `trained_submission.pkl` also saved in epoch_0/. Then use this model for epoch 1.

For me where we decide to put the `trained_submission.pkl` (and the corresponding `trained_agent.pkl` if any) is arbitrary, we can put it in `epoch_0` (because obtained from the trace located in the same folder) or put it in `epoch_1` because that’s the model (respectively the agent) used in epoch_1. I have no strong opinion for the moment.

**Comment by albertcthomas** [comment]
The `reward_function` or the `env` and even other parameters such as the number of epochs could go in `problem.py` but this would make `problem.py` more RL oriented while currently it is principally made for the static evaluations of the models. It's true that if you change the update schedule of the model this makes a different problem/benchmark maybe. We can go back to this later IMO.

**Comment by albertcthomas** [comment]
Two possibilities when running the `model-based-rl` script installed when installing the `rlkit` package. We can run it from the system folder or from outside a system folder specifying the system name in the command. I would tend toward the former, so that we run `ramp-test` and `model-based-rl` from the same folder.

**Comment by albertcthomas** [comment]
> `submissions/<submission>/<agent>/seed_<i>/epoch_<i>/(trained_submission.pkl, trace.csv`)

I would prefer `outputs/<submission>/<agent>/seed_<i>/epoch_<i>/(trained_submission.pkl, trace.csv)`

**Comment by kegl** [comment]
1. I think mbrlkit would be a better name than rlkit for the library. I'm also thinking about renaming this very repo to model_based_rl now that it's more than just a simulator.
2. Let's use data labels in `/data`, with a label that is also on option paramter to `generate_static_traces.py`.
3. I'm ok putting the first agent in `epoch_0`, I also don't have a strong opinion on it.
4. I agree to keep `problem.py` as is, and put rl parameters in another file (`mbrl_problem.py`?)
5. `ramp-test` has optional parameters allowing it to be run from anywhere (specifying `--ramp-kit-dir`, even `--ramp-submission-dir`), we could do the same so we can have flexibility where we run the script from.
6. I prefer quiet strongly to keep everything in `submissions/<submission>`, except that add a folder level like we do in `training_output`: `submissions/<submission>/mbrl_output/<agent>/seed_<i>/epoch_<i>/(trained_submission.pkl, trace.csv)`.
7. Where do we store the correspondance of `seed_<i>` and i?

**Comment by gabriel-hurtado** [comment]
1. I agree with @kegl on this point, and think renaming the repo itself makes sense
2. It would be nice indeed. If we do that we also need to modify train_test_split_initial_trace.py
3. I agree, first agent should be with epoch_0
4. Also think that problem.py should not be changed too much. Aggregating the rl related files to something like rl_spec.py is a good idea too

I have no strong opinion for 5 and 6.

7. The i refers to np.random.seed(i)

**Comment by albertcthomas** [comment]
>     1. I think mbrlkit would be a better name than rlkit for the library.

`mbrlkit` is indeed better. `mbrltools` is also an option. I am usning `mbrlkit` for now.


>     2\. Let's use data labels in `/data`, with a label that is also on option paramter to `generate_static_traces.py`.

This should be ok



>     5\. `ramp-test` has optional parameters allowing it to be run from anywhere (specifying `--ramp-kit-dir`, even `--ramp-submission-dir`), we could do the same so we can have flexibility where we run the script from.

Why not. however this would imply some changes such as the path of `metadata.json` in `problem.py`. `ramp-test` does not work outside the system folder with the current version of the code.

>     6\. I prefer quiet strongly to keep everything in `submissions/<submission>`, except that add a folder level like we do in `training_output`: `submissions/<submission>/mbrl_output/<agent>/seed_<i>/epoch_<i>/(trained_submission.pkl, trace.csv)`.

OK.



>     1. The i refers to np.random.seed(i)

Yes the `i` is the `i` of `--seed i`.

**Comment by albertcthomas** [comment]
It would be better to store the state of the RNG but we have different RNGs here (the one from `numpy` and the one from `pytorch`) and their could be other ones...

**Comment by kegl** [comment]
I would go for `mbrl-tools` with a dash.

**Comment by albertcthomas** [comment]
> I would go for `mbrl-tools` with a dash.

I like it for the name of the folder (as `ramp-workflow` or `scikit-learn`) but for the package we need something without a dash and preferably without underscores if we want to follow [PEP8 naming recommendation for packages](https://www.python.org/dev/peps/pep-0008/#package-and-module-names) (as `rampwf` or `sklearn`). I would go for `mbrl-tools` for the name of the folder and `mbrltools` for the name of the python package. I'll wait before doing the change, see if I have a better idea.

**Comment by kegl** [comment]
Yes, sounds good! `import mbrltools as mt`

**Comment by albertcthomas** [comment]
Closing now that everything is in place. We can open new issues for other discussions if needed.
